File: src/lib/awm.py
import json
from typing import List, Dict, Any, Optional
import asyncio
from openai import APIError, AsyncOpenAI, AuthenticationError, RateLimitError
import aiofiles
import backoff
from .workflow import Workflow, WorkflowStep
from .embedding_memory import EmbeddingMemory
import os

class AgentWorkflowMemory:
    def __init__(self, api_key: str, is_online_mode: bool = True, embedding_memory: Optional[EmbeddingMemory] = None):
        self.workflows: List[Workflow] = []
        self.openai = AsyncOpenAI(api_key=api_key)
        self.max_workflows: int = 100
        self.workflow_usage_history: List[str] = []
        self.is_online_mode = is_online_mode
        self.embedding_memory = embedding_memory

    async def induce_workflow(self, experiences: List[str]) -> None:
        try:
            if self.is_online_mode:
                await self.online_workflow_induction(experiences)
            else:
                await self.offline_workflow_induction(experiences)
            
            print(f"Number of workflows after induction: {len(self.workflows)}")
            for workflow in self.workflows:
                print(f"Workflow description: {workflow.description}")
                print(f"Number of steps: {len(workflow.steps)}")
        except AuthenticationError as auth_error:
            print('OpenAI API Authentication Error:', str(auth_error))
            print('Please check your API key and ensure it is valid.')
            await self.use_fallback_workflow_induction(experiences)
        except RateLimitError as rate_error:
            print('OpenAI API Rate Limit Error:', str(rate_error))
            await self.retry_with_backoff(lambda: self.induce_workflow(experiences))
        except Exception as error:
            print('Error inducing workflow:', str(error))
            await self.use_fallback_workflow_induction(experiences)

    async def online_workflow_induction(self, experiences: List[str]) -> None:
        prompt = f"""Given the following web navigation tasks, extract common workflows:

        Tasks:
        {chr(10).join(experiences)}

        For each workflow, provide:
        1. A brief description
        2. A list of steps, where each step includes:
           - Observation
           - Reasoning
           - Action

        Format the output as follows:

        ## Workflow Name
        Description: Brief description of the workflow
        Steps:
        1. Observation: What the agent observes
           Reasoning: Why the agent takes this action
           Action: The action taken

        2. Observation: ...
           Reasoning: ...
           Action: ...

        ## Next Workflow Name
        ..."""

        response = await self.openai.chat.completions.create(
            model="gpt-4",
            messages=[{"role": "user", "content": prompt}]
        )

        workflows_text = response.choices[0].message.content
        if workflows_text:
            self.parse_and_store_workflows(workflows_text)
        else:
            raise Exception('No workflow text generated')

    @backoff.on_exception(backoff.expo, Exception, max_tries=3)
    async def retry_with_backoff(self, fn):
        await fn()

    async def use_fallback_workflow_induction(self, experiences: List[str]) -> None:
        fallback_workflows = [
            Workflow(
                description=f"Fallback workflow for: {exp[:50]}...",
                steps=[WorkflowStep(
                    observation='Fallback observation',
                    reasoning='Using fallback due to API error',
                    action=exp
                )]
            ) for exp in experiences
        ]
        self.workflows.extend(fallback_workflows)

    async def offline_workflow_induction(self, experiences: List[str]) -> None:
        common_actions = ['click', 'type', 'select', 'submit']
        
        for exp in experiences:
            steps = []
            for action in exp.split('\n'):
                action_type = next((a for a in common_actions if a in action.lower()), 'perform')
                steps.append(WorkflowStep(
                    observation=f"Observed need to {action_type}",
                    reasoning="Action required based on task description",
                    action=action.strip()
                ))
            
            self.workflows.append(Workflow(
                description=f"Workflow induced offline: {exp[:50]}...",
                steps=steps
            ))

    def parse_and_store_workflows(self, workflows_text: str) -> None:
        workflows = [w.strip() for w in workflows_text.split('## ') if w.strip()]
        
        for workflow in workflows:
            lines = workflow.split('\n')
            name = lines[0].strip()
            description = next((line.replace('Description:', '').strip() for line in lines if line.startswith('Description:')), '')
            
            steps = []
            current_step = {}
            for line in lines[lines.index('Steps:') + 1:]:
                if line.startswith('Observation:'):
                    if current_step:
                        steps.append(WorkflowStep(**current_step))
                        current_step = {}
                    current_step['observation'] = line.replace('Observation:', '').strip()
                elif line.startswith('Reasoning:'):
                    current_step['reasoning'] = line.replace('Reasoning:', '').strip()
                elif line.startswith('Action:'):
                    current_step['action'] = line.replace('Action:', '').strip()
            
            if current_step:
                steps.append(WorkflowStep(**current_step))

            self.workflows.append(Workflow(description, steps))

    async def get_relevant_workflows(self, task: str) -> List[Workflow]:
        if self.is_online_mode and self.embedding_memory:
            context = await self.embedding_memory.get_relevant_context(task)
            return await self.select_workflows_with_context(task, context)
        else:
            return await self.select_workflows(task)

    async def select_workflows(self, task: str) -> List[Workflow]:
        # Simple keyword matching for offline mode
        keywords = task.lower().split()
        return [
            workflow for workflow in self.workflows
            if any(keyword in workflow.description.lower() for keyword in keywords)
        ]

    async def select_workflows_with_context(self, task: str, context: List[Any]) -> List[Workflow]:
        if not self.workflows:
            print("No workflows available for selection.")
            return []

        prompt = f"""Given the task: "{task}" and the following context:

        {json.dumps(context, indent=2)}

        Select the most relevant workflows from the following list:

        {json.dumps([{"index": i, "description": w.description} for i, w in enumerate(self.workflows)], indent=2)}

        Return the indices of the selected workflows as a JSON list."""

        response = await self.openai.chat.completions.create(
            model="gpt-4",
            messages=[{"role": "user", "content": prompt}]
        )

        try:
            content = response.choices[0].message.content.strip()
            # Check if the content is wrapped in backticks and remove them if present
            if content.startswith('```') and content.endswith('```'):
                content = content[3:-3].strip()
            selected_indices = json.loads(content)
            return [self.workflows[i] for i in selected_indices if 0 <= i < len(self.workflows)]
        except json.JSONDecodeError as e:
            print(f"Error decoding workflow selection response: {e}")
            print(f"Raw response: {response.choices[0].message.content}")
            # Fallback to simple keyword matching
            return await self.select_workflows(task)
        except Exception as e:
            print(f"Unexpected error in select_workflows_with_context: {e}")
            # Fallback to simple keyword matching
            return await self.select_workflows(task)

    async def generalize_workflow(self, workflow: Workflow):
        generalized = await self._apply_generalization_methods(workflow)
        await self.store_generalized_workflow(generalized)

    async def _apply_generalization_methods(self, workflow: Workflow) -> Workflow:
        workflow = await self._cross_task_generalization(workflow)
        workflow = await self._cross_website_generalization(workflow)
        workflow = await self._cross_domain_generalization(workflow)
        return workflow

    async def _cross_task_generalization(self, workflow: Workflow) -> Workflow:
        prompt = f"""Generalize the following workflow to make it applicable across different tasks:

        {json.dumps(workflow.__dict__, indent=2)}

        Provide the generalized workflow in the same JSON format."""

        response = await self.openai.chat.completions.create(
            model="gpt-4",
            messages=[{"role": "user", "content": prompt}]
        )

        try:
            generalized_data = json.loads(response.choices[0].message.content)
            return Workflow(**generalized_data)
        except json.JSONDecodeError:
            print("Error decoding cross-task generalization response")
            return workflow

    async def _cross_website_generalization(self, workflow: Workflow) -> Workflow:
        prompt = f"""Generalize the following workflow to make it applicable across different websites:

        {json.dumps(workflow.__dict__, indent=2)}

        Provide the generalized workflow in the same JSON format."""

        response = await self.openai.chat.completions.create(
            model="gpt-4",
            messages=[{"role": "user", "content": prompt}]
        )

        try:
            generalized_data = json.loads(response.choices[0].message.content)
            return Workflow(**generalized_data)
        except json.JSONDecodeError:
            print("Error decoding cross-website generalization response")
            return workflow

    async def _cross_domain_generalization(self, workflow: Workflow) -> Workflow:
        prompt = f"""Generalize the following workflow to make it applicable across different domains:

        {json.dumps(workflow.__dict__, indent=2)}

        Provide the generalized workflow in the same JSON format."""

        response = await self.openai.chat.completions.create(
            model="gpt-4",
            messages=[{"role": "user", "content": prompt}]
        )

        try:
            generalized_data = json.loads(response.choices[0].message.content)
            return Workflow(**generalized_data)
        except json.JSONDecodeError:
            print("Error decoding cross-domain generalization response")
            return workflow

    async def store_generalized_workflow(self, workflow: Workflow):
        if self.embedding_memory:
            await self.embedding_memory.store(workflow.description, workflow)
        if len(self.workflows) >= self.max_workflows:
            self.workflows.pop(0)  # Remove the oldest workflow
        self.workflows.append(workflow)

    async def evaluate_workflow(self, workflow: Workflow) -> dict:
        coverage = await self._calculate_coverage(workflow)
        function_overlap = await self._calculate_function_overlap(workflow)
        utility_rate = await self._calculate_utility_rate(workflow)

        return {
            "coverage": coverage,
            "function_overlap": function_overlap,
            "utility_rate": utility_rate,
        }

    async def _calculate_coverage(self, workflow: Workflow) -> float:
        # Calculate the percentage of common web tasks covered by the workflow
        common_tasks = ["navigation", "form filling", "data extraction", "interaction"]
        covered_tasks = sum(1 for task in common_tasks if any(task.lower() in step.action.lower() for step in workflow.steps))
        return covered_tasks / len(common_tasks)

    async def _calculate_function_overlap(self, workflow: Workflow) -> float:
        # Calculate the overlap of functionality with other workflows
        overlapping_steps = sum(1 for step in workflow.steps if any(step.action in other_workflow.steps for other_workflow in self.workflows if other_workflow != workflow))
        return overlapping_steps / len(workflow.steps) if workflow.steps else 0

    async def _calculate_utility_rate(self, workflow: Workflow) -> float:
        # Calculate the utility rate based on usage history
        workflow_uses = self.workflow_usage_history.count(workflow.description)
        total_uses = len(self.workflow_usage_history)
        return workflow_uses / total_uses if total_uses > 0 else 0

    async def update_workflows(self, task: str, solution: str) -> None:
        experience = f"{task}\n{solution}"
        await self.induce_workflow([experience])
        self.prune_workflows()
        await self.adapt_workflows(experience)

    async def adapt_workflows(self, experience: str) -> None:
        relevant_workflows = await self.get_relevant_workflows(experience)
        for workflow in relevant_workflows:
            self.update_workflow_steps(workflow, experience)
            self.update_workflow_variables(workflow, experience)

    def update_workflow_steps(self, workflow: Workflow, experience: str) -> None:
        experience_steps = experience.split('\n')
        for step in experience_steps:
            if not any(s.action == step for s in workflow.steps):
                workflow.steps.append(WorkflowStep(
                    observation='New step from experience',
                    reasoning='Added based on new experience',
                    action=step
                ))

    def update_workflow_variables(self, workflow: Workflow, experience: str) -> None:
        # Do nothing to preserve the original workflow
        pass

    def prune_workflows(self) -> None:
        if len(self.workflows) > self.max_workflows:
            sorted_workflows = sorted(self.workflows, 
                key=lambda w: self.workflow_usage_history.count(w.description), 
                reverse=True)
            self.workflows = sorted_workflows[:self.max_workflows]

    def track_workflow_usage(self, workflow_id: str) -> None:
        self.workflow_usage_history.append(workflow_id)
        if len(self.workflow_usage_history) > 100:
            self.workflow_usage_history = self.workflow_usage_history[-100:]

    def apply_workflow(self, workflow: Workflow, context: Dict[str, Any]) -> List[str]:
        # Return the original steps without any variable replacement
        return [step.action for step in workflow.steps]

    def apply_step(self, step: WorkflowStep, context: Dict[str, Any]) -> str:
        # Return the original action without any variable replacement
        return step.action

    async def save_workflows(self) -> None:
        try:
            workflows_data = [
                {
                    "description": w.description,
                    "steps": [
                        {
                            "observation": s.observation,
                            "reasoning": s.reasoning,
                            "action": s.action
                        } for s in w.steps
                    ]
                } for w in self.workflows
            ]
            async with aiofiles.open('workflows.json', 'w') as f:
                await f.write(json.dumps(workflows_data, indent=2))
        except Exception as error:
            print('Error saving workflows:', error)

    async def load_workflows(self) -> None:
        try:
            async with aiofiles.open('workflows.json', 'r') as f:
                data = await f.read()
            workflows_data = json.loads(data)
            self.workflows = [
                Workflow(
                    description=w["description"],
                    steps=[
                        WorkflowStep(
                            observation=s["observation"],
                            reasoning=s["reasoning"],
                            action=s["action"]
                        ) for s in w["steps"]
                    ]
                ) for w in workflows_data
            ]
        except FileNotFoundError:
            print('No existing workflows file found. Starting with an empty workflow list.')
            self.workflows = []
        except json.JSONDecodeError:
            print('Error decoding workflows file. Starting with an empty workflow list.')
            self.workflows = []
        except Exception as error:
            print('Error loading workflows:', error)
            self.workflows = []

    async def get_all_workflows(self) -> List[Workflow]:
        return self.workflows

    async def apply_workflow(self, workflow_id: int, context: dict) -> str:
        if workflow_id < 0 or workflow_id >= len(self.workflows):
            return f"Invalid workflow ID: {workflow_id}"
        workflow = self.workflows[workflow_id]
        applied_steps = [self.apply_step(step, context) for step in workflow.steps]
        return "\n".join(applied_steps)

    def apply_step(self, step: WorkflowStep, context: dict) -> str:
        # Return the original action without any variable replacement
        return step.action

async def main():
    api_key = os.getenv("OPENAI_API_KEY")
    if not api_key:
        raise ValueError("Please set the OPENAI_API_KEY environment variable")
    
    agent_memory = AgentWorkflowMemory(api_key)
    
    # Induce a workflow
    await agent_memory.induce_workflow(["Navigate to website", "Click login button", "Enter username", "Enter password", "Click submit"])
    
    # Get relevant workflows
    relevant_workflows = await agent_memory.get_relevant_workflows("Login to a website")
    
    # Apply a workflow
    if relevant_workflows:
        actions = agent_memory.apply_workflow(relevant_workflows[0], {"username": "user123", "password": "pass123"})
        print("Actions to take:", actions)
    
    # Save workflows
    await agent_memory.save_workflows()
    
    # Load workflows
    await agent_memory.load_workflows()
    
    # Analyze workflow quality
    quality_metrics = await agent_memory.evaluate_workflow(relevant_workflows[0])
    print("Workflow quality metrics:", quality_metrics)

if __name__ == "__main__":
    asyncio.run(main())

File: src/lib/awm.ts
import { OpenAI } from 'openai';
import { Workflow } from './types';
import fs from 'fs/promises';
import { backOff } from 'exponential-backoff';

interface WorkflowStep {
  observation: string;
  reasoning: string;
  action: string;
}

class AgentWorkflowMemory {
  private workflows: Workflow[] = [];
  private openai: OpenAI;
  private maxWorkflows: number = 100; // Limit the number of stored workflows
  private workflowUsageHistory: string[] = [];
  private isOnlineMode: boolean;

  constructor(apiKey: string, isOnlineMode: boolean = true) {
    this.openai = new OpenAI({ apiKey });
    this.isOnlineMode = isOnlineMode;
  }

  /**
   * Induces new workflows from given experiences.
   * @param experiences Array of experience strings to induce workflows from.
   */
  async induceWorkflow(experiences: string[]): Promise<void> {
    try {
      if (this.isOnlineMode) {
        await this.onlineWorkflowInduction(experiences);
      } else {
        await this.offlineWorkflowInduction(experiences);
      }
    } catch (error) {
      console.error('Error inducing workflow:', error);
      if (error instanceof OpenAI.APIError) {
        if (error.status === 429) {
          await this.retryWithBackoff(() => this.induceWorkflow(experiences));
        } else {
          await this.useFallbackWorkflowInduction(experiences);
        }
      } else {
        throw new Error('Failed to induce workflow: ' + (error as Error).message);
      }
    }
  }

  private async onlineWorkflowInduction(experiences: string[]): Promise<void> {
    const prompt = `Given the following web navigation tasks, extract common workflows:

    Tasks:
    ${experiences.join('\n')}

    For each workflow, provide:
    1. A brief description
    2. A list of steps, where each step includes:
       - Observation
       - Reasoning
       - Action

    Format the output as follows:

    ## Workflow Name
    Description: Brief description of the workflow
    Steps:
    1. Observation: What the agent observes
       Reasoning: Why the agent takes this action
       Action: The action taken

    2. Observation: ...
       Reasoning: ...
       Action: ...

    ## Next Workflow Name
    ...`;

    const response = await this.openai.chat.completions.create({
      model: 'gpt-4',
      messages: [{ role: 'user', content: prompt }],
    });

    const workflowsText = response.choices[0].message.content;
    if (workflowsText) {
      this.parseAndStoreWorkflows(workflowsText);
    } else {
      throw new Error('No workflow text generated');
    }
  }

  private async retryWithBackoff(fn: () => Promise<void>, maxRetries = 3): Promise<void> {
    await backOff(() => fn(), {
      numOfAttempts: maxRetries,
      startingDelay: 1000,
      timeMultiple: 2,
      maxDelay: 10000,
    });
  }

  private async useFallbackWorkflowInduction(experiences: string[]): Promise<void> {
    const fallbackWorkflows: Workflow[] = experiences.map(exp => ({
      description: `Fallback workflow for: ${exp.substring(0, 50)}...`,
      steps: [{
        observation: 'Fallback observation',
        reasoning: 'Using fallback due to API error',
        action: exp
      }]
    }));

    this.workflows.push(...fallbackWorkflows);
  }

  private async offlineWorkflowInduction(experiences: string[]): Promise<void> {
    // Implement a simple rule-based system for offline induction
    const commonActions = ['click', 'type', 'select', 'submit'];
    
    experiences.forEach(exp => {
      const steps: WorkflowStep[] = exp.split('\n').map(action => {
        const actionType = commonActions.find(a => action.toLowerCase().includes(a)) || 'perform';
        return {
          observation: `Observed need to ${actionType}`,
          reasoning: `Action required based on task description`,
          action: action.trim()
        };
      });

      this.workflows.push({
        description: `Workflow induced offline: ${exp.substring(0, 50)}...`,
        steps
      });
    });
  }

  private parseAndStoreWorkflows(workflowsText: string): void {
    const workflows = workflowsText.split('## ').filter(w => w.trim() !== '');
    
    workflows.forEach(workflow => {
      const [name, ...content] = workflow.split('\n');
      const descriptionMatch = content.join('\n').match(/Description: (.*)/);
      const description = descriptionMatch ? descriptionMatch[1].trim() : '';
      
      const stepsText = content.join('\n').split(/\d+\.\s/).filter(s => s.trim() !== '');
      const steps = stepsText.map(step => {
        const [observation, reasoning, action] = step.split('\n').filter(s => s.trim() !== '');
        return {
          observation: observation.replace('Observation:', '').trim(),
          reasoning: reasoning.replace('Reasoning:', '').trim(),
          action: action.replace('Action:', '').trim()
        };
      });

      this.workflows.push({
        description,
        steps
      });
    });
  }

  /**
   * Retrieves workflows relevant to the given task.
   * @param task The task to find relevant workflows for.
   * @returns An array of relevant Workflow objects.
   */
  async getRelevantWorkflows(task: string): Promise<Workflow[]> {
    try {
      if (!this.isOnlineMode) {
        return this.offlineRelevantWorkflows(task);
      }

      const prompt = `Given the following task and a list of workflows, determine which workflows are relevant to the task. Return the indices of the relevant workflows, separated by commas.

Task: ${task}

Workflows:
${this.workflows.map((w, i) => `${i}. ${w.description}`).join('\n')}

Relevant workflow indices:`;

      const response = await this.openai.chat.completions.create({
        model: 'gpt-3.5-turbo',
        messages: [{ role: 'user', content: prompt }],
      });

      const content = response.choices[0].message.content;
      if (content) {
        const relevantIndices = content.split(',').map(i => parseInt(i.trim()));
        return relevantIndices.map(i => this.workflows[i]).filter((w): w is Workflow => w !== undefined);
      }
    } catch (error) {
      console.error('Error getting relevant workflows:', error);
      return this.offlineRelevantWorkflows(task);
    }
    return [];
  }

  private offlineRelevantWorkflows(task: string): Workflow[] {
    // Simple keyword matching for offline mode
    const keywords = task.toLowerCase().split(' ');
    return this.workflows.filter(w => 
      keywords.some(keyword => w.description.toLowerCase().includes(keyword))
    );
  }

  /**
   * Updates workflows based on new task and solution.
   * @param task The task that was performed.
   * @param solution The solution that was found for the task.
   */
  async updateWorkflows(task: string, solution: string): Promise<void> {
    const experience = `${task}\n${solution}`;
    await this.induceWorkflow([experience]);
    this.pruneWorkflows();
    this.generalizeWorkflows();
    await this.adaptWorkflows(experience);
  }

  private async adaptWorkflows(experience: string): Promise<void> {
    const relevantWorkflows = await this.getRelevantWorkflows(experience);
    for (const workflow of relevantWorkflows) {
      this.updateWorkflowSteps(workflow, experience);
      this.updateWorkflowVariables(workflow, experience);
    }
  }

  private updateWorkflowSteps(workflow: Workflow, experience: string): void {
    const experienceSteps = experience.split('\n');
    experienceSteps.forEach(step => {
      if (!workflow.steps.some(s => s.action === step)) {
        workflow.steps.push({
          observation: 'New step from experience',
          reasoning: 'Added based on new experience',
          action: step
        });
      }
    });
  }

  private updateWorkflowVariables(workflow: Workflow, experience: string): void {
    // Do nothing to preserve the original workflow
  }

  evaluateWorkflowQuality(): {coverage: number, functionOverlap: number, utilityRate: number} {
    const coverage = this.calculateCoverage();
    const functionOverlap = this.calculateFunctionOverlap();
    const utilityRate = this.calculateUtilityRate();
    return {coverage, functionOverlap, utilityRate};
  }

  private pruneWorkflows(): void {
    if (this.workflows.length > this.maxWorkflows) {
      // Sort workflows by usage frequency
      const sortedWorkflows = this.workflows.sort((a, b) => {
        const aUsage = this.workflowUsageHistory.filter(w => w === a.description).length;
        const bUsage = this.workflowUsageHistory.filter(w => w === b.description).length;
        return bUsage - aUsage;
      });

      // Keep only the top maxWorkflows
      this.workflows = sortedWorkflows.slice(0, this.maxWorkflows);
    }
  }

  generalizeWorkflows(): void {
    this.crossTaskGeneralization();
    this.crossWebsiteGeneralization();
    this.crossDomainGeneralization();
  }

  private crossTaskGeneralization(): void {
    for (let i = 0; i < this.workflows.length; i++) {
      for (let j = i + 1; j < this.workflows.length; j++) {
        const generalizedWorkflow = this.attemptGeneralization(this.workflows[i], this.workflows[j]);
        if (generalizedWorkflow) {
          this.workflows.push(generalizedWorkflow);
        }
      }
    }
  }

  private crossWebsiteGeneralization(): void {
    const websitePatterns = this.identifyWebsitePatterns();
    for (const pattern of websitePatterns) {
      const generalizedWorkflow = this.createGeneralizedWebsiteWorkflow(pattern);
      if (generalizedWorkflow) {
        this.workflows.push(generalizedWorkflow);
      }
    }
  }

  private crossDomainGeneralization(): void {
    const domainGroups = this.groupWorkflowsByDomain();
    for (const domain in domainGroups) {
      const domainWorkflows = domainGroups[domain];
      const generalizedWorkflow = this.createGeneralizedDomainWorkflow(domainWorkflows);
      if (generalizedWorkflow) {
        this.workflows.push(generalizedWorkflow);
      }
    }
  }

  private identifyWebsitePatterns(): string[] {
    const patterns: string[] = [];
    const commonActions = ['login', 'search', 'add to cart', 'checkout'];
    
    commonActions.forEach(action => {
      const matchingWorkflows = this.workflows.filter(w => 
        w.steps.some(s => s.action.toLowerCase().includes(action))
      );
      if (matchingWorkflows.length > 1) {
        patterns.push(action);
      }
    });

    return patterns;
  }

  private createGeneralizedWebsiteWorkflow(pattern: string): Workflow | null {
    const matchingWorkflows = this.workflows.filter(w => 
      w.steps.some(s => s.action.toLowerCase().includes(pattern))
    );

    if (matchingWorkflows.length < 2) return null;

    const generalizedSteps: WorkflowStep[] = [];
    const maxSteps = Math.max(...matchingWorkflows.map(w => w.steps.length));

    for (let i = 0; i < maxSteps; i++) {
      const stepActions = matchingWorkflows.map(w => w.steps[i]?.action || '');
      const generalizedAction = this.generalizeText(...stepActions);
      generalizedSteps.push({
        observation: `Generalized observation for ${pattern} step ${i + 1}`,
        reasoning: `Common step across ${pattern} workflows`,
        action: generalizedAction
      });
    }

    return {
      description: `Generalized ${pattern} workflow`,
      steps: generalizedSteps
    };
  }

  private groupWorkflowsByDomain(): Record<string, Workflow[]> {
    const domainGroups: Record<string, Workflow[]> = {};
    
    this.workflows.forEach(workflow => {
      const domain = this.extractDomain(workflow.description);
      if (!domainGroups[domain]) {
        domainGroups[domain] = [];
      }
      domainGroups[domain].push(workflow);
    });

    return domainGroups;
  }

  private extractDomain(description: string): string {
    const domains = ['e-commerce', 'social media', 'productivity', 'finance'];
    return domains.find(d => description.toLowerCase().includes(d)) || 'general';
  }

  private createGeneralizedDomainWorkflow(domainWorkflows: Workflow[]): Workflow | null {
    if (domainWorkflows.length < 2) return null;

    const generalizedSteps: WorkflowStep[] = [];
    const maxSteps = Math.max(...domainWorkflows.map(w => w.steps.length));

    for (let i = 0; i < maxSteps; i++) {
      const stepActions = domainWorkflows.map(w => w.steps[i]?.action || '');
      const generalizedAction = this.generalizeText(...stepActions);
      generalizedSteps.push({
        observation: `Generalized domain observation for step ${i + 1}`,
        reasoning: `Common step across domain workflows`,
        action: generalizedAction
      });
    }

    return {
      description: `Generalized workflow for ${this.extractDomain(domainWorkflows[0].description)}`,
      steps: generalizedSteps
    };
  }

  private attemptGeneralization(w1: Workflow, w2: Workflow): Workflow | null {
    if (w1.steps.length !== w2.steps.length) return null;

    const generalizedSteps = w1.steps.map((step, index) => {
      const step2 = w2.steps[index];
      return {
        observation: this.generalizeText(step.observation, step2.observation),
        reasoning: this.generalizeText(step.reasoning, step2.reasoning),
        action: this.generalizeText(step.action, step2.action),
      };
    });

    return {
      description: `Generalized: ${w1.description} / ${w2.description}`,
      steps: generalizedSteps,
    };
  }

  private generalizeText(...texts: string[]): string {
    const words = texts.map(text => text.split(' '));
    const generalizedWords = words[0].map((word, index) => {
      if (words.every(w => w[index] === word)) {
        return word;
      }
      return '{variable}';
    });
    return generalizedWords.join(' ');
  }

  analyzeWorkflowQuality(): {coverage: number, functionOverlap: number, utilityRate: number} {
    const coverage = this.calculateCoverage();
    const functionOverlap = this.calculateFunctionOverlap();
    const utilityRate = this.calculateUtilityRate();
    return {coverage, functionOverlap, utilityRate};
  }

  private calculateWorkflowQuality(workflow: Workflow): number {
    // Implement a scoring system based on workflow characteristics
    const stepCount = workflow.steps.length;
    const uniqueActions = new Set(workflow.steps.map(s => s.action)).size;
    return stepCount * uniqueActions; // Simple quality score
  }

  private calculateCoverage(): number {
    const totalActions = this.workflows.reduce((sum, w) => sum + w.steps.length, 0);
    const uniqueActions = new Set(this.workflows.flatMap(w => w.steps.map(s => s.action))).size;
    return totalActions > 0 ? uniqueActions / totalActions : 0;
  }

  private calculateFunctionOverlap(): number {
    const allActions = this.workflows.flatMap(w => w.steps.map(s => s.action));
    const actionCounts = allActions.reduce((acc, action) => {
      acc[action] = (acc[action] || 0) + 1;
      return acc;
    }, {} as Record<string, number>);
    const overlapCount = Object.values(actionCounts).filter(count => count > 1).length;
    return Object.keys(actionCounts).length > 0 ? overlapCount / Object.keys(actionCounts).length : 0;
  }

  private calculateUtilityRate(): number {
    if (this.workflowUsageHistory.length === 0) {
      return 0; // No usage data available
    }

    const totalWorkflows = this.workflows.length;
    const usedWorkflows = new Set(this.workflowUsageHistory).size;
    const usageRate = usedWorkflows / totalWorkflows;

    const recentUsageWeight = 0.7;
    const totalUsageWeight = 0.3;

    const recentUsageCount = this.workflowUsageHistory.slice(-10).length;
    const recentUsageRate = recentUsageCount / 10;

    const weightedUtilityRate = (recentUsageRate * recentUsageWeight) + (usageRate * totalUsageWeight);

    return weightedUtilityRate;
  }

  private trackWorkflowUsage(workflowId: string): void {
    this.workflowUsageHistory.push(workflowId);
    // Limit history to last 100 uses to prevent unbounded growth
    if (this.workflowUsageHistory.length > 100) {
      this.workflowUsageHistory = this.workflowUsageHistory.slice(-100);
    }
  }

  /**
   * Applies a workflow to the given context and returns an array of actions.
   * @param workflow The workflow to apply.
   * @param context The context to apply the workflow to.
   * @returns An array of actions derived from the workflow.
   */
  applyWorkflow(workflow: Workflow, context: Record<string, any>): string[] {
    // Return the original steps without any variable replacement
    return workflow.steps.map(step => step.action);
  }

  private applyStep(step: WorkflowStep, context: Record<string, any>): string {
    // Return the original action without any variable replacement
    return step.action;
  }

  /**
   * Saves the workflows to a file.
   */
  async saveWorkflows(): Promise<void> {
    try {
      await fs.writeFile('workflows.json', JSON.stringify(this.workflows));
    } catch (error) {
      console.error('Error saving workflows:', error);
    }
  }

  /**
   * Loads the workflows from a file.
   */
  async loadWorkflows(): Promise<void> {
    try {
      const data = await fs.readFile('workflows.json', 'utf-8');
      this.workflows = JSON.parse(data);
    } catch (error) {
      console.error('Error loading workflows:', error);
      // If file doesn't exist or is corrupted, start with an empty workflow list
      this.workflows = [];
    }
  }
}

export default AgentWorkflowMemory;